### DISCUSSION TITLE
Can I run glsl directly?

### DISCUSSION THREAD

**Discussion by psionic12** [gfx-rs/wgpu#1812] [Q&A]
I'm trying to port my engine to rust, and all my shaders are written in glsl. I know I can compile them into spri-v, but at this time of development, anyway I can run a glsl directly?

**Comment by kvark** [comment]
The shading language of WebGPU is WGSL: https://gpuweb.github.io/gpuweb/wgsl/
We also support SPIR-V behind "spirv" feature, but I wouldn't recommend going this path if you can avoid it.
What you can do is using [naga](https://github.com/gfx-rs/naga) to convert GLSL to WGSL, either at build time or at run-time (depending on your app requirements), and feed that WGSL to wgpu.
You can quickly test how well it works (or doesn't) by checking out naga and doing this:
```
cargo run -- <your_file.glsl> <new_file.wgsl>
```

**Reply by psionic12** [reply]
Thanks for helping, I'll try this naga to check if all my shaders can be convert into WGSL well.
However I'm not very clear about how WGSL is compiled. Does it compiles into spri-v by wgpu, or somehow feed the source code into the driver directly (let's say wgpu use vulkan as the backend) ?

**Reply by kvark** [reply]
`wgpu` uses `naga` to compile WGSL to the backend shader language, internally. On Vulkan that would be SPIR-V.

**Comment by Atlas16A** [comment]
I know this is an old post but its relevant to what im doing, Im currently making a game in bevy and I need to do some ray tracing for some advanced topology math, i found a way to do it in a glsl compute shader but as far as I know wgpu still doesn't have support for ray tracing so im not sure if I can even convert this to wgsl without things breaking, so wondering If i can use glsl directly to get around what wgpu lacks.

**Reply by cwfitzgerald** [reply]
Is it using hardware raytracing (aka RayQuery) or just standard compute raytracing - if it's the latter, it'll work.

To a small extent you can use "spirv-passthrough" to pass through raw shaders, but you also need to enable the vulkan features, which is unfortunately not trivial.

**Reply by kvark** [reply]
I did land Ray Query support in Naga, which I've been using in https://github.com/kvark/blade. It's not a part of WGSL standard though.
However, there would still be needed a lot of host-side boilerplate for state binding and object creation, which I don't think wgpu or Bevy is currently offering.

**Reply by cwfitzgerald** [reply]
It's landed in wgpu-hal over vk, not yet in core

**Comment by nevakrien** [comment]
I seen that there is a glsl feature flag on the main page but I can't find docs as to how it works

**Reply by cwfitzgerald** [reply]
Enable the feature flag, hand us the [glsl variant] (https://docs.rs/wgpu/latest/wgpu/enum.ShaderSource.html#variant.Glsl.field.shader) of shader source.
